### PR TITLE
Fixed several issues related to company-lsp--post-completion

### PR DESCRIPTION
**Pull request by yyoncho**
1. Fixed issue related to expanding snippets after additionalEdits have been
applied. The snipped was expanded on wrong position becase applying
additionalEdits may move the points.

2. Fixed applying text edits when the completion has been triggered by completion
character. The old code was keeping the text in company-lsp--line-backup but if
you trigger the completion with . it was not part of company-lsp--line-backup.
As a result company-lsp was deleting the last character you have written.
Now the region to delete is calulated based on the inserted candidate which
works similar to company--insert-candidate.

3. When there is textEdit, the insert text should be the newText from the
textEdit.

### CODE REVIEW

File: company-lsp.el [hunk 1]
```diff
@@ -235,27 +232,31 @@ CANDIDATE is a string returned by `company-lsp--make-candidate'."
...
-      (lsp--apply-text-edit text-edit)
+      (setq insert-text (gethash "newText" text-edit))
+      (delete-region (- (point) (length candidate)) (point))
       (let* ((range (gethash "range" text-edit))
              (start-point (lsp--position-to-point (gethash "start" range)))
-             (new-text-length (length (gethash "newText" text-edit))))
+             (new-text-length (length insert-text)))
+        ;; set end of the range to be equal to the start of the range
+        ;; since we have already deleted the prefix.
+        (puthash "end" (gethash "start" range) range)
```
Comment (tigersoldier) [inline company-lsp.el]: This assumes that the text edit removes the prefix and inserts the whole completion candidate. There is no guarantee the server always sends text edit in this fashion

File: company-lsp.el [hunk 2]
```diff
@@ -235,27 +232,31 @@ CANDIDATE is a string returned by `company-lsp--make-candidate'."
...
-    (when additional-text-edits
-      (lsp--apply-text-edits additional-text-edits))
-    (when (and company-lsp-enable-snippet
-               (fboundp 'yas-expand-snippet))
-      (if (and insert-text (eq insert-text-format 2))
-          (yas-expand-snippet insert-text start (point))
-        (-when-let (fallback-snippet (company-lsp--fallback-snippet item))
-          (yas-expand-snippet fallback-snippet))))
+
+    (let ((start-marker (set-marker (make-marker) start)))
```
Comment (tigersoldier) [inline company-lsp.el]: Should clean up the marker at the end of the `let`:

```elisp
(set-marker start-marker nil)
```

**Comment by tigersoldier** [comment]
BTW I had #37 that is targeted to fix it, but didn't submitted to master because I didn't have time testing it recently. Can you try it and see if it works?

File: company-lsp.el (hunk 2 above)
Comment (yyoncho) [reply company-lsp.el]: Thanks, sure! I was using markers for the first time :)

**Comment by yyoncho** [comment]
#37 wont work since it will delete the completion char as well. E. g. if you have prefix Foo and you have FooBar selected in the company dropdown you may complete the sequence using dot or whatever you have configured in company-auto-complete-chars .

**Comment by tigersoldier** [comment]
Looks good. Thanks!

**Pull request merged**